imap_processing/ultra/l1b/ultra_l1b_extended.py

```diff
@@ -315,3 +319,96 @@ def get_ssd_back_position_and_tof_offset(
         ] = get_image_params(f"TOFSSDRTOFF{i}")
 
     return yb, tof_offset, ssd_number
+
+
+def get_coincidence_positions(
+    de_dataset: xarray.Dataset, particle_tof: np.ndarray, sensor: str
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate coincidence positions.
+
+    Calculate time for electrons to travel back to
+    the coincidence anode (etof) and the x coincidence position (xc).
+
+    The tof measured by the coincidence anode consists of the particle
+    tof from start to stop, plus the time for the electrons to travel
+    back to the coincidence anode.
+
+    Further description is available on pages 34-35 of
+    IMAP-Ultra Flight Software Specification document
+    (7523-9009_Rev_-.pdf).
+
+    Parameters
+    ----------
+    de_dataset : xarray.Dataset
+        Data in xarray format.
+    particle_tof : np.ndarray
+        Particle time of flight (i.e. from start to stop)
+        (tenths of a nanosecond).
+    sensor : str
+        Sensor name.
+
+    Returns
+    -------
+    etof : np.ndarray
+        Time for the electrons to travel back to
+        coincidence anode (tenths of a nanosecond).
+    xc : np.ndarray
+        X coincidence position (hundredths of a millimeter).
+    """
+    index_top = np.nonzero(np.isin(de_dataset["COIN_TYPE"], CoinType.Top.value))[0]
+    de_top = de_dataset.isel(epoch=index_top)
```

Comment:
I don't know xarray, but this seems like an unintuitive way to get a filtered dataset. I get the first line, but don't understand why `de_dataset.isel(epoch=index_top)` is used. I would think that the indices can be used directly?

Might be worth looking at `xarray.Dataset.filter_by_attrs()`: https://docs.xarray.dev/en/stable/generated/xarray.Dataset.filter_by_attrs.html

Reply:
I think the incantation might be:
```python
de_dataset.sel(COIN_TYPE=CoinType.Top.value)
```

`sel` is used for selecting against values, and `isel` for the indices. You have the values above, so you can just do the comparison within `sel`.

Reply:
Thank you for the suggestion! I understand that sel() is generally preferred for selecting data based on values, and I’ve considered using it here. However, since COIN_TYPE is a data variable and not a coordinate, I can't directly use sel(). 

 I’m opting to use np.nonzero() for indexing..